File: c/fermat.c
#include <gmp.h>

void fermat_naive(mpz_t n, mpz_t a, mpz_t b) {
        
    mpz_t SQRT, r;

    mpz_inits(SQRT, r);
    mpz_sqrtrem(SQRT, r, n);

    mpz_t u, v;
    mpz_inits(u, v);

    mpz_mul_ui(u, SQRT, 2);
    mpz_add_ui(u, u, 1);
    
    mpz_set_ui(v, 1);
    
    while (mpz_sgn(r) != 0){
        
        if (mpz_sgn(r) > 0){
            while (mpz_sgn(r) > 0){
                mpz_sub(r, r, v);
                mpz_add_ui(v, v, 2);
            }
        }

        if (mpz_sgn(r) < 0){
            mpz_add(r, r, u);
            mpz_add_ui(u, u, 2);
        }
    
    }

    mpz_add(a, u, v);
    mpz_sub_ui(a, a, 2);
    mpz_divexact_ui(a, a, 2);

    mpz_sub(b, u, v);
    mpz_divexact_ui(b, b, 2);
}

File: _python_deprecated/fermat.py
from gmpy2 import mpz, isqrt_rem

def fermat_naive(n):
    n = mpz(n)
        
    SQRT, r = isqrt_rem(n)
    r *= -1
    
    u = 2 * SQRT + 1
    v = 1
    
    while r != 0:
        
        if r > 0:
            while r > 0:
                r -= v
                v += 2
        
        if r < 0:
            r += u
            u += 2
    
    a = (u + v -2)/2
    b = (u - v)/2
    
    print(f"n = {a} * {b}")